src/test/util/setup_common.cpp

```diff
@@ -87,6 +88,7 @@ BasicTestingSetup::BasicTestingSetup(const std::string& chainName, const std::ve
         extra_args);
     util::ThreadRename("test");
     fs::create_directories(m_path_root);
+    m_args.ForceSetArg("-datadir", m_path_root.string());
```

Comment:
There is already m_node.args. Any reason to avoid using that?

Reply:
> There is already m_node.args. Any reason to avoid using that?

I suggested this for a followup (It would be a bigger change involving init and might also require updating more tests). Diff was here: https://github.com/bitcoin/bitcoin/pull/21244#issuecomment-816026224

Reply:
Until then I'd prefer to use `m_node.args->` (or `gArgs.` if that is too much typing). It will be a trivial scripted diff to change that to `m_args.`. But having three argsman in the tests (two of which are the same and one that is different) will just cause confusion and makes review harder/code more fragile. How would you protect against their state diverging?

Reply:
I'm happy with any intermediate state. Here's what I think the ultimate goal should be:

- Tests should have an `ArgsManager m_args` member that is newly initialized before each test and completely destroyed after each test.
- Where possible, tests should pass `m_args` references directly to functions that are being tested and need to read options.
- Tests can pass `m_args` pointers indirectly through `NodeContext` when they are calling node RPC methods, or through `WalletContext` when they are calling wallet RPC methods.
- Unless tests are calling RPC methods or init code, they probably don't need `NodeContext` and `WalletContext` instances and it would be nice to drop `NodeContext` from the base test class.
- `gArgs` should go away in node and wallet code. It might be useful to keep `gArgs` in test code to read test options (random seeds, timeouts, log options, debug options). It might make sense to keep `gArgs` in GUI code since GUI code relies on other global singletons like `QApplications` and `QSettings` and trying to get rid of it might just add extra complexity.

The current PR seems just like a good a first step in this direction to me since it adds local test `ArgsManager` instances and starts to use them minimally.

These are just my thoughts though. I think any approach you want to take here is probably reasonable.

Reply:
Fair enough, any approach fine by me, too

Reply:
I think Russell summarized it very nicely. 

My approach with this PR is to use `m_args` wrt `GetDataDir()` functionality for tests. Yes, it is somewhat fragile at this point. If you are too concerned about this, I can revert that change in `BasicTestingSetup` and leave it for some follow-up PR. 

note: There are tests that just need `GetDataDir()` function and not anything else from `ArgsManager`, these tests benefit from using `m_args` (instead of `gArgs`).

note 2: Originally, https://github.com/bitcoin/bitcoin/commit/6e7e67f855282ae4baa3756f2e95aee12de13f91 was a part of this PR and with this, one can see quite nicely where `gArgs` is used and that's one step closer to remove it from those places and make some more tests to be true unit tests.